File: packages/core/src/pagination.ts
/**
 * Pagination behavior machinery.
 * Contract: docs/contracts/components/pagination.md, "Behavior Machine".
 *
 * Pure page-window math and navigation guards; the adapter owns controller
 * integration, scrolling, and callbacks.
 */

export type VisiblePage = number | "ellipsis";

/**
 * Page-number window: first and last pages always visible, `siblings` pages
 * around the current one, ellipsis markers for gaps of more than one page.
 */
export function buildVisiblePages(page: number, count: number, siblings: number): VisiblePage[] {
  const pages = new Set<number>([1, count]);

  for (let candidate = page - siblings; candidate <= page + siblings; candidate += 1) {
    if (candidate >= 1 && candidate <= count) {
      pages.add(candidate);
    }
  }

  const sorted = Array.from(pages).sort((left, right) => left - right);
  const result: VisiblePage[] = [];

  for (let index = 0; index < sorted.length; index += 1) {
    const current = sorted[index];
    const previous = sorted[index - 1];

    if (current === undefined) {
      continue;
    }

    if (previous !== undefined && current - previous > 1) {
      result.push("ellipsis");
    }

    result.push(current);
  }

  return result;
}

/** A page request is valid when in bounds and actually a navigation. */
export function canRequestPage(nextPage: number, currentPage: number, totalPages: number): boolean {
  return Number.isInteger(nextPage) && nextPage >= 1 && nextPage <= totalPages && nextPage !== currentPage;
}

File: packages/contracts/headless/src/pagination.rs
//! Pagination machinery. Mirror of core `pagination.ts`.

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum VisiblePage {
    Page(usize),
    Ellipsis,
}

/// Page-number window: first and last pages always visible, `siblings`
/// pages around the current one, ellipsis markers for gaps of more than one
/// page. Exact mirror of the TS `buildVisiblePages` (set-based).
pub fn build_visible_pages(page: usize, count: usize, siblings: usize) -> Vec<VisiblePage> {
    if count == 0 {
        return vec![];
    }

    let mut pages: Vec<usize> = vec![1, count];

    let low = page.saturating_sub(siblings).max(1);
    let high = (page + siblings).min(count);

    for candidate in low..=high {
        pages.push(candidate);
    }

    pages.sort_unstable();
    pages.dedup();

    let mut result = Vec::new();

    for (index, current) in pages.iter().enumerate() {
        if index > 0 && current - pages[index - 1] > 1 {
            result.push(VisiblePage::Ellipsis);
        }

        result.push(VisiblePage::Page(*current));
    }

    result
}

/// A page request is valid when in bounds and actually a navigation.
pub fn can_request_page(next_page: i64, current_page: i64, total_pages: i64) -> bool {
    next_page >= 1 && next_page <= total_pages && next_page != current_page
}
